skrf/vi/vna/vna.py

```diff
@@ -0,0 +1,739 @@
+            self.write_binary = self.resource.write_binary_values
+
+            self.query = self.resource.query
+            self.query_ascii = self.resource.query_ascii_values
+            self.query_binary = self.resource.query_binary_values
+
+        else:
+            raise NotImplementedError(
+                "Only message based resources are implemented at this time."
+            )
+
+        if isinstance(self.resource, pyvisa.resources.GPIBInstrument):
+            self.resource.control_ren(RENLineOperation.deassert_gtl)
+
+    def __enter__(self) -> VNA:
+        """
+        context manager entry point
+
+        Returns
+        -------
+        VNA
+            the Analyzer Driver Object
+        """
+        return self
+
+    def __exit__(self, *args, **kwargs) -> None:
+        """
+        context manager exit point
+
+        Parameters
+        ----------
+        *args
+        **kwargs
+        """
+        self.resource.close()
+
+    def close(self) -> None:
+        """
+        Close the device connection
+        """
+        self.__exit__(None, None, None)
+
+    def reset(self) -> None:
+        """
+        Reset the device to default
+        """
+        self.write("*RST")
+
+    def id_string(self) -> str:
+        """
+        Query the device identifier string
+
+        Returns
+        -------
+        str
+            device identifier string
+        """
+        return self.query("*IDN?")
+
+    def start_freq(self, channel: int = 1) -> float:
+        """
+        Get start frequency
+
+        Parameters
+        ---------
+        channel : int
+            channel to get frequency (if supported)
+
+        Returns
+        -------
+        float
+            start frequency [Hz]
+        """
+        return float(self.query(f"sense{channel}:frequency:start?"))
+
+    def set_start_freq(self, f: float, channel: int = 1) -> None:
+        """
+        Set start frequency
+
+        Parameters
+        ----------
+        f : float
+            start frequency [Hz]
+        channel : int
+            channel to set frequency (if supported)
+        """
+        self.write(f"sense{channel}:frequency:start {f}")
+
+    def stop_freq(self, channel: int = 1) -> float:
+        """
+        Get stop frequency
+
+        Parameters
+        ----------
+        channel : int
+            channel to get frequency (if supported)
+
+        Returns
+        -------
+        float
+            stop frequency [Hz]
+        """
+        return float(self.query(f"sense{channel}:frequency:stop?"))
+
+    def set_stop_freq(self, f: float, channel: int = 1) -> None:
+        """
+        Set stop frequency
+
+        Parameters
+        ----------
+        f : float
+            stop frequency [Hz]
+        channel : int
+            channel to set frequency (if supported)
+        """
+        self.write(f"sense{channel}:frequency:stop {f}")
+
+    def npoints(self, channel: int = 1) -> int:
```

Comment:
Open question for my knowledge : are they some VNA supporting having different points and frequency for each channel ?

i was wondering if the `channel` parameter was of interest (because if not, most these parameters could be implemented as `@property` and `.setter`...

Reply:
From what I've been able to gather, some instruments have multiple channels (e.g. Keysight PNA, Rhode & Schwarz ZVA) and others don't (Keysight FieldFox, NanoVNA)

As for what a "channel" I've hard a hard time figuring out or if it's even in the SCPI standard. I had initially started writing hoping we could do properties but made the switch when I saw it was more than just the keysight that used channels. 

After Vince's comments I thought about it more and I think we can still do this with properties but define them like this

```python
@start_freq.setter
def start_frequency(
    self, 
    freq: Union[float, str, Tuple[int, Union[float, str]]
) -> None:
    if isinstance(freq, tuple):
        # set with channel
    elif isinstance (freq, str):
        # convert freq string to float and set with default channel
    elif isinstance (freq, float):
        # set using default channel
    else:
        raise RuntimeError("Invalid argument")

```

In the base class, the branch could just raise NotImplementedError and those instruments that support channels could implement. 

Thinking about it more, should we also create a UnsupportedError to better reflect when an instrument doesn't have some functionality?

Reply:
> I think we can still do this with properties but define them like this

yes eventually. To be tested if it simplifies the API. 

I would raise a [ValueError](https://docs.python.org/3/library/exceptions.html#ValueError) in case of invalid argument.

how does it work on the VNA you've tested? Is `channel` was always 1?

Reply:
So I've tested both the PNA (specifically the [Keysight N5227B](https://www.keysight.com/us/en/support/N5227B/pna-microwave-network-analyzer-900-hz-10-mhz-67-ghz.html)) and the [FieldFox](https://www.keysight.com/us/en/products/network-analyzers/fieldfox-handheld-rf-microwave-analyzers.html). The FieldFox doesn't have channels at all, but for the PNA, the [their docs](http://web.archive.org/web/20220521165946/https://na.support.keysight.com/pna/help/latest/S0_Start/Traces_Channels_and_Windows.htm#ManagingChan) are helpful:

> Channels contain traces. The analyzer can have up to 200 independent channels.
>
> Channel settings determine how the trace data is measured . All traces that are assigned to a channel share the same channel settings. A channel must be selected (active) to modify its settings. To select a channel, click the [Trace Status](http://web.archive.org/web/20220521165946/https://na.support.keysight.com/pna/help/latest/S1_Settings/Customize_Your_Analyzer_Screen.htm#trace_status) button of a Trace in that channel. The following are channel settings:
> * Frequency Range
> * Power level
> * Calibration
> * IF Bandwidth
> * Number of Points
> * Sweep Settings
> * Average
> * Trigger (some settings are global)

Now I'm not sure if this explanation is consistent across manufacturers. This is the language from the [R&S ZVA's manual](https://www.rohde-schwarz.com/webhelp/zva_html_usermanual_en/zva_html_usermanual_en.htm)

> A channel contains hardware-related settings to specify how the network analyzer collects data. The channel settings can be divided into three main groups:
>
>    * Control of the measurement process (Sweep, Trigger, Average)
>
>    * Description of the test setup (Power of the internal source, IF filter Bandwidth and Step Attenuators, Port Configuration)
>
>    * Correction data (Calibration, Offset)

When I've done testing, I've really only used channel 1, but I will test with other channels and report back / push another commit.

Reply:
@jhillairet I'm working on switching to properties, but already hit a snag. With query functions like "start_freq", they can take an argument for the channel. Right now I'm leaning toward something like:

```python
    @property
    def start_freq(self) -> Union[float, Tuple[int, float]]:
        """
        Get start frequency on the **active** channel

        Returns
        -------
        float or (int, float)
            start frequency [Hz]. If the instrument supports channels, returns a
            tuple of the active channel and it's start frequency
        """
        # if instrument has channel support:
            channel = self.active_channel
            return (channel, float(self.query(f"sense{channel}:frequency:start?")))
        # else:
            return float(self.query(f"sense{channel}:frequency:start?"))
```

Which means the setter would have to be different from what's described above, as it would then only make sense to make the setter change the settings of the active channel. That would then mean each time we call the setter, we'd have to query the instrument for the active channel.

```python
    @property
    def active_channel(self) -> int:
        # if instrument doesn't support channels:
            return 1
        # else:
            # I think this is instrument specific
    ...
    @start_freq.setter
    def start_freq(self, freq: float) -> None:
        channel = self.active_channel
        self.write(f"sense{channel}:frequency:start {freq}")
```

Reply:
Hey @dvincentwest. Yeah SCPI is quite a loose standard which is frustrating at times. The comment above was more just about deciding whether or not to use properties for a lot of things instead of get/set functions and how to do that. 

I do think providing a unified high-level API for VNAs is worthwhile though. There are a certain set of things that every user of a VNA is trying to do at a minumum that I do think should be consistent across devices:

 - configure a measurement (frequency, power, etc)
 - calibrate
 - get data out

I don't think the effort here precludes any specific subclass of VNA from implementing commands specific to their particular feature set. Again though, here I am thinking through implementation and wonder if doing something like mixins would be a way to get around details like channels. Just have something like `VNAWithChannelsMixin` or `VNAWithoutChannelsMixin`. Maybe that's getting ahead of myself 😅 

On doing stuff with GUIs, a huge motivation for me doing this refactor was in my work on another project, [PyChamber](https://github.com/HRG-Lab/PyChamber) (for anechoic measurements) and I think would be beneficial for other GUI applications as well!

Finally, I haven't taken any of your comments as discouragement. Your experience is super valuable and I appreciate your thoughts!

Reply:
@cafeclimber that certainly adds some context and if you have a serious effort to integrate multiple VNAs into GUI code then you do need to have some kind of standard APIs for that.  As an aside, I wish PyChamber had existed 6 years ago when I was doing Anechoic chamber measurements.  The documentation for the positioners was almost non-existent, and what I left behind was basically golden-path code that no one understands.  Again, best of luck with that.  From here I don't think I have much left to contribute.  I see I am listed as a reviewer on this pull request, but I don't think it would be appropriate for me to gate your work here given how long I have been detached from the code.

Reply:
@dvincentwest  thanks for commenting in on this.  i also no longer have VNA access which makes working on this hard, but i think having good VI  support (specially VNAs) is a really important feature.